### ISSUE TITLE
Native selection mode: some Buttons (including menu buttons, expandable buttons etc.) are not copied

### ISSUE THREAD

**Issue by Adriani90** [nvaccess/nvda#15910]
### Steps to reproduce:
1. With NVDA running, enable native selection mode in Firefox
2. For example on this Github page at the top side, there are some buttons, i.e.:
"Button, Type / to search "
"Button, Command palette"
"Menu button, Create something new"
Select them with shift+arrow keys and try to copy these.

### Actual behavior:
NVDA does not copy anything, in some cases nVDA copies symbols such as bullet points but not the label of the buttons.

### Expected behavior:
NVDA should copy the visible label of the buttons if any, or an embeded object showing the visible form of the object / placeholder to the clipboard, same as it happens when using the mouse.

### NVDA logs, crash dumps and other attachments:
N/A

#### NVDA installed/portable/running from source:
installed

#### NVDA version:
alpha-30299,4519b893 (2024.1.0.30299)

#### Windows version:
Windows 11 23 H2

#### Name and version of other software in use when reproducing the issue:
Firefox 120

#### Other information about your system:
n/a

#### Does the issue still occur after restarting your computer?
yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
n/a

#### If NVDA add-ons are disabled, is your problem still occurring?
yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
yes

**Comment by Adriani90** [comment]
This happens in other situations as well, for example when copying the label of tabs on this Github page such as
"tab, write selected"
"tab, Preview"

**Comment by brunoprietog** [comment]
I imagine this is because what NVDA reads are the aria-labels of these buttons, but visually there is no text

**Comment by Adriani90** [comment]
Good point, thanks for this info.

In this case I think it is useful to at least document in the user guide what exactly is natively copied in such cases (i.e. unicode characters or the visible symbols themselves?

cc: @CyrilleB79 what do you think?

**Comment by CyrilleB79** [comment]
As @XLTechie pointed out very interestingly in https://github.com/nvaccess/nvda/issues/15909#issuecomment-1851595634:
> My understanding is that you should get the same as you would with NVDA turned off.

What happens when NVDA is turned off (or when pressing NVDA+F2 then control+C)? I have not tested.

**Comment by seanbudd** [comment]
Cc @jcsteh - this is a firefox issue

**Comment by Adriani90** [comment]
It seems when copying tabs with NVDA turned off, there are embeded objects copied in MS Word but I am not sure what they look like. @Qchristensen could you help with some vision here?
I attached here a document with some embeded objects that are copied when NVDA is turned off but you cannot copy them with NVDA on and the native selection enabled which are the tabs "write" and "preview" if I understand this properly.

I am not really able to copy buttons without visible labels because it is really hard for me to select them with the mouse without sight.
but at least for buttons with visible labels it seems the text is copied as expected both with NVDA native selection and with the mouse.
[Embeded elements.docx](https://github.com/nvaccess/nvda/files/13654055/Embeded.elements.docx)

**Comment by jcsteh** [comment]
If I copy the tabs, I do get "Write", "Preview", etc. in my clipboard.

One challenge here is that when you select a button or similar to copy, Firefox has no way of knowing whether you want just the content of that element or also the element itself. If I select the Write tab, for example, I only get the "Write" text, but not the semantic fact that it's a tab. I don't really know how we can solve this. There can be many ancestor elements and there's no reasonable way of specifying which ones you want to include. A sighted user will have a similar experience. Most things you could paste into won't be able to handle interactive semantics anyway.

**Comment by jcsteh** [comment]
Regarding nothing at all being copied in some cases,  I suspect this is the same underlying cause as #15909. I would expect something to be copied, even if it's not quite what you expect. For example, if it's a button with an aria-label and no text content, it might be an empty button which you can't usefully paste into most applications.

**Comment by Adriani90** [comment]
Thanks for this info Jamie, actually I have trouble to get the text "write" or "preview" into my clipboard when using native selection.
But actually that's ok because when I copy these tabs with the mouse without NVDA, there are embeded objects being pasted in MS Word without the text as shown in the attached document. however, these embeded objects should also appear when i copy them with the native selection mode in NVDA.

Re buttons without visible labels / text, actually I would expect that also such embeded objects are copied to the clipboard, or at least any kind of symbol, but maybe @Qchristensen or @seanbudd can provide a feedback on what is actually copied when using the mouse on such buttons or elements without visible text.

**Comment by Qchristensen** [comment]
If I copy the top of this page, it actually looks pretty similar using the mouse without NVDA and using native selection with NVDA.  Where I found a difference is if I copy using the mouse in Chrome.

In all cases, the headers:
Code
Issues 2.4k
Pull requests 38
Discussions
Actions
Projects 7

were all copied, as links.  In Chrome, buttons (such as Back to notifications, done, unsubscribe, mark as unread and save - or if not coming from notifications the close and comment buttons down the bottom) get copied as text.  In firefox these are skipped.

It's noticeable yes, but perhaps taking a step back towards @jcsteh's question - what is the use case?  In most cases, the reason for copying (formatted) information from web pages is to copy text so it looks similar, headings are larger, links are links, etc, and images come across too.  In every instance I have ever wanted to copy from a web page into say Word or email or wherever, I haven't wanted / tried to capture buttons - which is not to say there isn't a use case for it, just it's not as obvious - and if wanting more than just the text, buttons are more likely to do things beyond a simple link which you wouldn't be able to recreate outside the browser in any case.

Having said that, I can see the argument for wanting the text at least, but I'll leave it to others to work out the technicality of that.

**Comment by Adriani90** [comment]
@Qchristensen thanks for your input here.
Actually I was not interested in having the semantic of a button copied to the clipboard, but the visual placeholder. I don't have enough sight to assess, but if you look at the document I have attached in https://github.com/nvaccess/nvda/issues/15910#issuecomment-1852958304, there are some embeded objects in there. NVDA says "embeded object" and nothing else. This was when I used the mouse to copy the tabs "tab write" and "tab preview" in the comment area on this github page.
I don't get these embeded objects to the clipboard when using native selection.

How are these embeded objects in the attached document actually look like?

I have also edited the issue description maybe it is clearer now.

**Comment by Adriani90** [comment]
@jcsteh, @michaelDCurran I think I could narrow this issue down as follows. It seems automatically moving focus to focusable elements plays an important role here, or at least it is crucial that users synchronize them manually with NVDA+spacebar before starting the selection. This needs clarification in the user guide I think:
* Buttons with visible symbols but no visible text label are not copied to clipboard, at least I could not succeed to copy the placeholders with the mouse either. So it is not about the semantics, but about the visible part on the screen (e.g. the symbol visibly shown). Should I create a separate issue for this?

### Regarding selecting tabs:
1. Go to browse mode
2. Make sure your virtual cursor is not synchronized with the system focus
3. Put the virtual cursor on this github page on the "write tab"
4. Press shift+down arrow key twice to select "write tab" and "preview tab"
5. Press ctrl+c
6. Open an MS Word document and paste the content from the clipboard
7. Syncrhonize system focus to the virtual cursor on the tab "write" or "preview" with NVDA+spacebar
8. Repeat steps 3-6
9. Do another test starting from "write tab" but press shift+down arrow key three times, to also select "headings button" which is the first item of the toolbox for text formating.

### Actual:
a. Step 2+ Steps 3-6: if the system focus is not manually synchronized to the tab area "e.g. to the write tab via nvda+space bar", NVDA does not copy anything to the clipboard
b1. Step 7 + steps 3-6: If the system focus is synchronized to the "write tab" or preview tab via NVDA+space bar, then native selection selects the first tab only when pressing shift+down arrow twice while starting on "write tab".
This being said, in case of b1. we can copy each tab natively to the clipboard, but each of it needs to be selected separately. Selecting both at the same time with shift+down arrow will not put both of them to the clipboard.
b2. steps 7 + steps 3-6 variation: After pressing shift+down arrow twice, press shift+left arrow to unselect the space character rendered in browse mode between the tab area and the toolbox for text formating: in this case, NVDA will copy both tabs "write tab" and "preview tab" to the clipboard as expected. So it seems the space character rendered in browse mode has a direct effect on native selection.
c. Step 7 + step 9: native selection fails and nothing is copied to the clipboard, because we extend the selection over the tab area up to the first item of the text formating toolbox, and the system focus is only synchronized to the tab area. So it doesn't move to the toolbox area.

### Expected:
a. Clarification in the user guide about the selection process, e.g. system focus needs to be synchronized manually via NVDA+space bar before starting selection of focusable elements such as links, tabs or anything else with tabindex.
b1, b2 and c: Pressing shift+down arrow should select and copy natively elements in diferent regions of the page (e.g. tab area and toolbox). There should not be the need to unselect any separators or space characters for this to work properly.

**Comment by jcsteh** [comment]
There are improvements to this functionality in Firefox 140 (currently Nightly), so some of this may be addressed.

**Comment by Adriani90** [comment]
@jcsteh in Firefox 140 this is definitely more consistent than before. Thanks for the great work.
It seems the semantics of headings are not copied. However, this seems to happen in general, in every browser also when selecting headings with the mouse and copy them. Only plain text is comming out of the clipboard when pasting it in MS Word. So it's not related to NVDA.

Closing this as invalid, because the buttons and tabs with aria-label I was talking about in the original description, including tabs, are not copied on mouse selection either. Before Firefox 140, these were appearing as embeded objects in MS Word when pasting them after mouse selection and copy action. This was not working when selecting them with browse mode caret in native selection mode.
